### PR TITLE
scylla_util.py - Azure Instance Checks

### PR DESCRIPTION
**Pull request by fee-mendes**
scylla_io_setup auto-tuning will always fail when running on top of Azure. This commit fixes:

- [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=linux) address 
- A contradiction (i.e: `if False and True:`) when checking whether the instance is supported

I have decided to keep `is_unsupported_instance_class` method intact for future use, although it currently always returns `False`.

### CODE REVIEW

File: dist/common/scripts/scylla_util.py [hunk 1]
```diff
@@ -380,9 +380,9 @@ def __init__(self):
 
     @staticmethod
     def is_azure_instance():
-        """Check if it's Azure instance via DNS lookup to metadata server."""
+        """Check if it's Azure instance via DNS lookup to IMDS server."""
         try:
-            addrlist = socket.getaddrinfo('metadata.azure.internal', 80)
```
Comment (avikivity) [inline dist/common/scripts/scylla_util.py:385]: What's wrong with that? Please add the explanation to the patch changelog.

Comment (yaronkaikov) [reply dist/common/scripts/scylla_util.py:385]: @tarzanek the url is probably missing from /etc/hosts file. we had the same discusstion in https://github.com/scylladb/scylla/pull/8596

Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:385]: but your change is merged @yaronkaikov, so how come this is still text and not octets?

Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:385]: ah, you reverted, ok, I see now, well ... so is this Azure bug then? but if Felipe hit it and you hit it Yaron, then we should likely switch to IP

Comment (tarzanek) [reply dist/common/scripts/scylla_util.py:385]: unless you know a way to repro and we can report to Azure, but despite that we should work in both cases

**Review by tarzanek — CHANGES_REQUESTED** [review]
as Avi said, please document in the message why we have to go for IP in check if it is azure instance

Also this can have possible negative effects in the way where you run the same code on GCP ... since metadata server IP is the same for both, but hostname isn't (and that does the distinguish)
So please check if we detect GCP BEFORE we detect Azure, so we don't get false positives from this test
Also @yaronkaikov and @syuu1228 might be using this call in other parts of scylla_setup and again we need to make sure it won't generate false positives (so GCP will be detected as Azure)

File: dist/common/scripts/scylla_util.py (hunk 1 above)
Comment (fee-mendes) [reply dist/common/scripts/scylla_util.py:385]: I couldn't find any references to metadata.azure.internal. Not sure where this came out from. 

Azure IMDS docs has no mentions to `metadata.azure.internal`. Wasn't this a simply copy'n'paste from GCP class? :-) 

I'll update the description for clarity.

File: dist/common/scripts/scylla_util.py [hunk 2]
```diff
@@ -381,14 +384,13 @@ def __init__(self):
     @staticmethod
     def is_azure_instance():
         """Check if it's Azure instance via DNS lookup to metadata server."""
+        p = azure_instance()
```
Comment (tarzanek) [inline dist/common/scripts/scylla_util.py:387]: creating instance might trigger code you don't want to run on a machine that is not azure
remove and rather make later pointers static

File: dist/common/scripts/scylla_util.py [hunk 3]
```diff
@@ -381,14 +384,13 @@ def __init__(self):
...
         try:
-            addrlist = socket.getaddrinfo('metadata.azure.internal', 80)
-        except socket.gaierror:
+            is_imds = curl(p.META_DATA_BASE_URL + p.API_VERSION, headers = { "Metadata": "True" }, byte=False, timeout=3, max_retries=2, retry_interval=1)
```
Comment (tarzanek) [inline dist/common/scripts/scylla_util.py:389]: so this means on an on premise machine
this code will wait 2 + 2 x 3 seconds just to continue and say it's not Azure

in such case you can just go resolve 169.254.169.254

**Comment by mykaul** [comment]
@fee-mendes - is this PR still relevant?

**Comment by fee-mendes** [comment]
I dont remember it, so no, closed.

**Pull request closed without merging**